### ISSUE TITLE
Reachable objects are reclaimed when using GC_DEBUG

### ISSUE THREAD

**Issue by hamadrana99** [ivmai/bdwgc#280]
When I do not define GC_DEBUG before #include <gc.h>, my program runs as expected. But I want to debug and print backtraces, so I define GC_DEBUG. In gdb I put a breakpoint at GC_finish_collection(), when it breaks, I print info about a certain object. The debugger shows me that it is reachable by printing a trace when I do print GC_print_backtrace(object address) and shows me that it is marked as 1 when I do GC_is_marked(object address).

(gdb) p GC_print_backtrace(0x7ffff7e2a2b0)
0x7ffff7e2a2b0 (../examples/resnet152-imagenet.c:72, sz=16, NORMAL)
	Call chain at allocation:
		/home/ahmed90/bdwgc/.libs/libgc.so.1(+0x17fae) [0x7fffeb1ebfae]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x95255) [0x5555555e9255]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0xa0ca8) [0x5555555f4ca8]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x8adaf) [0x5555555dedaf]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x83be5) [0x5555555d7be5]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
Reachable via 0 levels of pointers from offset 32 in object:
0x7ffff7ec3a70 (../examples/resnet152-imagenet.c:72, sz=40, NORMAL)
	Call chain at allocation:
		/home/ahmed90/bdwgc/.libs/libgc.so.1(+0x17fae) [0x7fffeb1ebfae]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x156da8) [0x5555556aada8]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x83be5) [0x5555555d7be5]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
Reachable via 1 levels of pointers from offset 24 in object:
0x7ffff7ec3930 (../examples/resnet152-imagenet.c:72, sz=32, NORMAL)
	Call chain at allocation:
		/home/ahmed90/bdwgc/.libs/libgc.so.1(+0x17fae) [0x7fffeb1ebfae]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x203abc) [0x555555757abc]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x83be5) [0x5555555d7be5]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
Reachable via 2 levels of pointers from offset 32 in object:
0x7ffff7ec36b0 (../examples/resnet152-imagenet.c:72, sz=40, NORMAL)
	Call chain at allocation:
		/home/ahmed90/bdwgc/.libs/libgc.so.1(+0x17fae) [0x7fffeb1ebfae]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0xee358) [0x555555642358]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x83be5) [0x5555555d7be5]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
Reachable via 3 levels of pointers from offset 24 in object:
0x7ffff7ec34d0 (../examples/resnet152-imagenet.c:72, sz=32, NORMAL)
	Call chain at allocation:
		/home/ahmed90/bdwgc/.libs/libgc.so.1(+0x17fae) [0x7fffeb1ebfae]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x1ce07c) [0x55555572207c]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x83be5) [0x5555555d7be5]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
		/home/ahmed90/checkpoint-vlad/examples/resnet152-imagenet(+0x9b680) [0x5555555ef680]
Reachable via 4 levels of pointers from root at 0x7fffffffc800

(gdb) p GC_is_marked(GC_base(0x7ffff7e2a2b0))
$3 = 1

But when then I do

(gdb) p GC_dump_finalization()
Disappearing (short) links:
Disappearing long links:
Finalizers:
Finalizable object: 0x7ffff7e2a2b0
Finalizable object: 0x7ffff7e79e80
$8 = void

It shows that the object is finalizable. And I hit continue and that object gets collected instantly. Although there is references to it as shown by the print backtrace.

**Comment by ivmai** [comment]
Finalizable object is normally marked, so it could be collected only next GC cycle after the finalizer is executed.

**Comment by ivmai** [comment]
Is the object got reclaimed even there's a reference to it?

**Comment by hamadrana99** [comment]
Yes that’s what i am saying. The object should be reachable. And the backtrace also says that it is reachable. But it gets collected. And this only happens if I have GC_DEBUG. If I omit it then it works fine.

**Comment by ivmai** [comment]
I assume you are using the latest stable release e.g. v8.0.4 (or master).
It's hard to me to understand where's the bug. Seems that some pointer is not observed by GC because of displacement (needed for debug header).
If GC_all_interior_pointers == 0 then try with GC_all_interior_pointers=1.

**Comment by hamadrana99** [comment]
I cloned it from github and built it with --enable-gc-debug. I have tried setting GC_all_interior_pointers to both 0 and 1. It gives me a warning that it is deprecated, setting it to 1 exhibits the same behavior, setting it to 0 segfaults the program at the first ever call to GC_malloc.

If you can give me any pointers to what you think is causing this i.e. enabling GC_DEBUG causes it to collect everything even though the backtrace shows that it is reachable (I posted this stack in my original post), I will try to take it from there and investigate.
________________________________

**Comment by hboehm** [comment]
Does the backtrace it gives you originate at a real root? Is 0x7fffffffc800
in the original message the root that you thought was holding onto the
object? Or is it just telling you that the object is now reachable from the
finalization queue, since it is now enqueued for finalization, but it was
never really reachable?
